### DISCUSSION TITLE
Duplicate database object with pg-promise

### DISCUSSION THREAD

**Discussion by PaulKushch** [vercel/next.js#18008] [Help]
# Bug report
Getting "WARNING: Creating a duplicate database object for the same connection." with pg-promise

## Describe the bug
It seems that in development mode in next js the warning is present while the connection best practices are followed, outlined by pg-promise author at https://stackoverflow.com/questions/34382796/where-should-i-initialize-pg-promise.

Because Node.js caches a module’s exported values, this should theoretically make every API route use the same db connection object. In Next.js app in development, each time the change is made to the source code, and webpack rebuilt and hot-deployed the assets, I get a warning:

**WARNING: Creating a duplicate database object for the same connection.**

To overcome this I resorted to using the Singleton pattern to re-use the database connection object.

## To Reproduce

I use the singleton pattern as outlined (and slightly modified by me) at https://www.codeoftheprogrammer.com/2020/01/16/postgresql-from-nextjs-api-route/
```
const connectionString = `postgresql://${process.env.DB_USER}:${process.env.DB_PASSWORD}@${process.env.DB_HOST}:${process.env.DB_PORT}/${process.env.DB_DB}`;
const cn = {
	capSQL: true,
};

const dbCn = {
	connectionString,
	max: 1,
};

// Use a symbol to store a global instance of a connection, and to access it from the singleton.
const DB_KEY = Symbol.for("MyApp.db");
const PGP_KEY = Symbol.for("MyApp.pgp");
const globalSymbols = Object.getOwnPropertySymbols(global);
const hasDb = globalSymbols.indexOf(DB_KEY) > -1;
if (!hasDb) {
	global[PGP_KEY] = require("pg-promise")(cn);
	global[DB_KEY] = global[PGP_KEY](dbCn);
}

// Create and freeze the singleton object so that it has an instance property.
const singleton = {};
Object.defineProperty(singleton, "db", {
	get() {
		return global[DB_KEY];
	},
});
Object.defineProperty(singleton, "pgp", {
	get() {
		return global[PGP_KEY];
	},
});
Object.freeze(singleton);

module.exports = {
	pgp: singleton.pgp,
	db: singleton.db,
};
```

## Expected behavior

I would expect that next js does not present that warning.
Maybe someone else got the same warning and has a better solution?
Best Regards

**Comment by willieseabrook** [comment]
Linking with https://github.com/vitaly-t/pg-promise/issues/175 for people who find this.
